### PR TITLE
Move Logs options to component-base/config

### LINKED ISSUE #99267: Move Logs options to component-base/config
#### What would you like to be added:

Migrate https://github.com/kubernetes/kubernetes/tree/master/staging/src/k8s.io/component-base/logs/options.go to https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/component-base/config

/kind cleanup

#### Why is this needed:

Configuration of logs is split between Options in component-base/logs/options.go used for control plane components and LoggingConfiguration in component-base/config/types.go. Having two flows for initialization flags can lead to them going out of sync leading to bugs.

Part of work for Structured Logging.

Previous attempt https://github.com/kubernetes/kubernetes/pull/92732
/sig instrumentation
/cc @ehashman 
/priority important-soon
/help

### PR DESCRIPTION
**Pull request by mengjiao-liu**
#### What type of PR is this?
/kind cleanup

#### What this PR does / why we need it:
Configuration of logs is split between Options in component-base/logs/options.go used for control plane components and LoggingConfiguration in component-base/config/types.go. Having two flows for initialization flags can lead to them going out of sync leading to bugs.

#### Which issue(s) this PR fixes:

Fixes #99267

#### Does this PR introduce a user-facing change?

```release-note
None
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by liggitt** [comment]
The API bits of this look to be limited to a couple constants. Once the structured logging aspects are reviewed, re-tag for api-review and I can ack those bits.

/remove-label api-review

File: staging/src/k8s.io/component-base/config/options/loggingconfig.go [hunk 1]
```diff
@@ -0,0 +1,66 @@
+package options
```
Comment (yuzhiquan) [inline staging/src/k8s.io/component-base/config/options/loggingconfig.go:1]: Need Copyright here.

File: staging/src/k8s.io/component-base/config/options/loggingconfig.go [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+
+import (
+	"flag"
+	"fmt"
+	"github.com/spf13/pflag"
+	"k8s.io/component-base/config"
+	"k8s.io/component-base/logs"
+	logs2 "k8s.io/component-base/logs/json"
+	"k8s.io/klog/v2"
+	"strings"
```
Comment (yuzhiquan) [inline staging/src/k8s.io/component-base/config/options/loggingconfig.go:11]: Should split by standard lib or third party lib.

File: staging/src/k8s.io/component-base/config/options/loggingconfig.go [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+	JsonLogFormat     = "json"
+	logFormatFlagName = "logging-format"
+)
+
+var LogRegistry = logs.NewLogFormatRegistry()
+
+func init() {
+	// Text format is default klog format
+	LogRegistry.Register(DefaultLogFormat, nil)
+	LogRegistry.Register(JsonLogFormat, logs2.JSONLogger)
```
Comment (yuzhiquan) [inline staging/src/k8s.io/component-base/config/options/loggingconfig.go:25]: Why `logs2`, how about `json` or `jsonlog`?

File: staging/src/k8s.io/component-base/logs/logs.go [hunk 4]
```diff
@@ -27,9 +27,9 @@ import (
 	"k8s.io/klog/v2"
 )
 
-const logFlushFreqFlagName = "log-flush-frequency"
+const LogFlushFreqFlagName = "log-flush-frequency"
```
Comment (yuzhiquan) [inline staging/src/k8s.io/component-base/logs/logs.go:30]: Why `logFlushFreqFlagName` to `LogFlushFreqFlagName`? Some other pkg import it?

File: staging/src/k8s.io/component-base/config/validation/validation.go [hunk 5]
```diff
@@ -17,8 +17,14 @@ limitations under the License.
...
 
 import (
+	"flag"
+	"fmt"
+	"github.com/spf13/pflag"
 	"k8s.io/apimachinery/pkg/util/validation/field"
 	"k8s.io/component-base/config"
+	"k8s.io/component-base/config/options"
+	"k8s.io/component-base/logs"
+	"strings"
```
Comment (yuzhiquan) [inline staging/src/k8s.io/component-base/config/validation/validation.go:27]: sort import

File: staging/src/k8s.io/component-base/config/options/loggingconfig.go (hunk 1 above)
Comment (mengjiao-liu) [reply staging/src/k8s.io/component-base/config/options/loggingconfig.go:1]: Thank you very much. I forgot about it

File: staging/src/k8s.io/component-base/logs/logs.go (hunk 4 above)
Comment (mengjiao-liu) [reply staging/src/k8s.io/component-base/logs/logs.go:30]: [staging/src/k8s.io/component-base/config/validation/validation.go:89](https://github.com/kubernetes/kubernetes/pull/99693/files#diff-5d0601b20c5ccb8c7abf87e7e0dd477eeeb6862cb186c96595c156aced23a5c8R89) uses it

File: staging/src/k8s.io/component-base/config/validation/validation.go (hunk 5 above)
Comment (mengjiao-liu) [reply staging/src/k8s.io/component-base/config/validation/validation.go:27]: I Should split by standard lib or third party lib. But the sorting is correct, and the goimport tool is used to sort it automatically

File: staging/src/k8s.io/component-base/config/options/loggingconfig.go [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+	"github.com/spf13/pflag"
+	"k8s.io/component-base/config"
+	"k8s.io/component-base/logs"
+	jsonlogs "k8s.io/component-base/logs/json"
+	"k8s.io/klog/v2"
+)
+
+const (
+	DefaultLogFormat  = "text"
+	JsonLogFormat     = "json"
```
Comment (yuzhiquan) [inline staging/src/k8s.io/component-base/config/options/loggingconfig.go:33]: s/JsonLogFormat/JSONLogFormat/

Comment (mengjiao-liu) [reply staging/src/k8s.io/component-base/config/options/loggingconfig.go]: Thanks. I've fixed it

File: staging/src/k8s.io/component-base/logs/options/options.go [hunk 7]
```diff
@@ -0,0 +1,63 @@
+/*
```
Comment (serathius) [inline staging/src/k8s.io/component-base/logs/options/options.go:1]: Is there any benefit from having this file in `staging/src/k8s.io/component-base/logs/options/options.go` instead of `staging/src/k8s.io/component-base/logs/options.go`. I just ask as this is less important to change of moving config and introduces changes in a lot of directories with different owners. 

I think we should split this move to make review and collecting approvals easier.

**Comment by DangerOnTheRanger** [comment]
Merge conflict aside, LGTM!

**Comment by mengjiao-liu** [comment]
> Overall great work!!!
> 
> Sorry for code conflict, we had some bug that need to be fixed before the cut (#99257). Could you rebase this PR?

Thanks for your review and I have rebased it

File: cmd/kubeadm/app/kubeadm.go [hunk 8]
```diff
@@ -30,7 +30,8 @@ import (
 
 // Run creates and executes new kubeadm command
 func Run() error {
-	klog.InitFlags(nil)
+	local := flag.NewFlagSet(os.Args[0], flag.ExitOnError)
+	klog.InitFlags(local)
```
Comment (neolit123) [inline cmd/kubeadm/app/kubeadm.go:34]: could you explain why this change is needed?

Comment (mengjiao-liu) [reply cmd/kubeadm/app/kubeadm.go:34]: Because the code moves and imports `k8s.io/component-base/log` package, the package has the init() function, which contains init log flag：
https://github.com/kubernetes/kubernetes/blob/a3806d6e497f5e235df9ae9157f29414a62dd3b4/staging/src/k8s.io/component-base/logs/logs.go#L34-L36

Leaving this line of code unchanged leads to `flag refined`

or do you have any better suggestions ? @neolit123

Comment (neolit123) [reply cmd/kubeadm/app/kubeadm.go:34]: > Because the log code change and kubeadm run imports k8s.io/component-base/log package

i cannot find direct imports of `k8s.io/component-base/log` in `/cmd/kubeadm`. does it get imported indirectly?

> Leaving this line of code klog.InitFlags(nil) unchanged leads to flag refined error

but this only happens after this PR, correct?

> the package has the init() function, which contains init log flag：

having something like:
```
func init() { 
 	klog.InitFlags(flag.CommandLine) 
 }
```

doesn't seem like a good idea in general.
this makes klog flags to be force added to the default command line for anyone who importers the `k8s.io/component-base/log` package even if they only want its utilities.

Comment (neolit123) [reply cmd/kubeadm/app/kubeadm.go:34]: if something else is doing the `klog.InitFlags` can we remove the `klog.InitFlags` call in kubeadm completely and replace it with a comment in the lines of:

> // package k8s.io/component-base/log is responsible for doing klog.InitFlags()

and add a `_ k8s.io/component-base/log` import on top?

Comment (mengjiao-liu) [reply cmd/kubeadm/app/kubeadm.go]: > if something else is doing the `klog.InitFlags` can we remove the `klog.InitFlags` call in kubeadm completely and replace it with a comment in the lines of:
> 
> > // package k8s.io/component-base/log is responsible for doing klog.InitFlags()
> 
> and add a `_ "k8s.io/component-base/log"` import on top?

Thanks, It looks better and clearer and It's easier to understand

Comment (mengjiao-liu) [reply cmd/kubeadm/app/kubeadm.go]: > i cannot find direct imports of k8s.io/component-base/log in /cmd/kubeadm. does it get imported indirectly?

it is indirect, and the following is the compilation package order：

```
package k8s.io/kubernetes/cmd/kubeadm/app
	imports k8s.io/kubernetes/cmd/kubeadm/app/cmd
	imports k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm/validation
	imports k8s.io/kubernetes/cmd/kubeadm/app/componentconfigs
	imports k8s.io/kube-proxy/config/v1alpha1
	imports k8s.io/component-base/config/v1alpha1
	imports k8s.io/component-base/config/options
	imports k8s.io/component-base/logs
```
> doesn't seem like a good idea in general.
this makes klog flags to be force added to the default command line for anyone who imports the k8s.io/component-base/log package even if they only want its utilities.

so I use change following code instead of klog.InitFlags(nil) to avoid using logs init()
```
	local := flag.NewFlagSet(os.Args[0], flag.ExitOnError)
	klog.InitFlags(local)
```
> but this only happens after this PR, correct?

This PR changes the following code and the following code uses k8s.io/component-base/config/options package,
https://github.com/kubernetes/kubernetes/pull/99693/files#diff-c0a54d42bf7c11099aaec980abd8ee871dfba9428da8b3ce8b8d480ef4d2989eR110-R113

k8s.io/component-base/config/options package import k8s.io/component-base/logs and causesk8s.io/component-base/logs init() function to be called

File: cmd/kubeadm/app/kubeadm.go [hunk 9]
```diff
@@ -22,15 +22,14 @@ import (
 
 	"github.com/spf13/pflag"
 
-	"k8s.io/klog/v2"
-
 	cliflag "k8s.io/component-base/cli/flag"
 	"k8s.io/kubernetes/cmd/kubeadm/app/cmd"
+	// package k8s.io/component-base/log is responsible for doing klog.InitFlags()
```
Comment (serathius) [inline cmd/kubeadm/app/kubeadm.go:27]: This is not true, currently code for registering klog flags is located in 
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/component-base/cli/globalflag/globalflags.go

As Kubeadm doesn't use either `k8s.io/component-base/cli` nor uses `k8s.io/component-base/logs` let's leave their klog invocation as it is.

Comment (mengjiao-liu) [reply cmd/kubeadm/app/kubeadm.go:27]: Indirectly import the package  `imports k8s.io/component-base/logs`
```
package k8s.io/kubernetes/cmd/kubeadm/app
	imports k8s.io/kubernetes/cmd/kubeadm/app/cmd
	imports k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm/validation
	imports k8s.io/kubernetes/cmd/kubeadm/app/componentconfigs
	imports k8s.io/kube-proxy/config/v1alpha1
	imports k8s.io/component-base/config/v1alpha1
	imports k8s.io/component-base/config/options
	imports k8s.io/component-base/logs
```
If the code does not change, an error is reported
https://github.com/kubernetes/kubernetes/pull/99693#discussion_r592033489

Is that what you're saying ?
```
	local := flag.NewFlagSet(os.Args[0], flag.ExitOnError)
	klog.InitFlags(local)
```

Comment (neolit123) [reply cmd/kubeadm/app/kubeadm.go:27]: > 	imports k8s.io/kube-proxy/config/v1alpha1
>	imports k8s.io/component-base/config/v1alpha1
>	imports k8s.io/component-base/config/options
>	imports k8s.io/component-base/logs

this is a problematic import that is the result of this PR and we need to resolve it.

this means that all consumers of the public kube-proxy API type or the `config/v1alpha1` type directly, will get their default CommandLine mutated...

we have to either make `component-base/config/v1alpha1` not use the `logs` package, or make the `logs` package not call `InitFlags()` in `init()`.

/hold

File: cmd/kubeadm/app/kubeadm.go (hunk 8 above)
Comment (liggitt) [reply cmd/kubeadm/app/kubeadm.go]: > having something like:
> 
> ```
> func init() { 
>  	klog.InitFlags(flag.CommandLine) 
>  }
> ```
> 
> doesn't seem like a good idea in general.
> this makes klog flags to be force added to the default command line for anyone who imports the `k8s.io/component-base/log` package even if they only want its utilities.

Strongly agree with this. packages that have side effects when you import them should not contain other things that would give people reasons to import them. As an example, the `install` packages for our API types (e.g. `pkg/apis/core/install/install.go`) are standalone packages, rather than installing in the package that defines the API type.

Comment (mengjiao-liu) [reply cmd/kubeadm/app/kubeadm.go]: I agree with you. 
I think it's better to delete this log package init() function

File: cmd/kubeadm/app/kubeadm.go (hunk 9 above)
Comment (mengjiao-liu) [reply cmd/kubeadm/app/kubeadm.go:27]: Thank you very much for your advice. I'm more inclined to delete this init method. Because it's really difficult to find out which package to import log package and call init function...

Comment (neolit123) [reply cmd/kubeadm/app/kubeadm.go:27]: i think it's worth investigating when it was added (probably near the klog v2 migration?) and see what can break if it's removed.

**Comment by ehashman** [comment]
/priority important-longterm

Let's wait until after thaw to merge this.

File: cmd/kubeadm/app/kubeadm.go (hunk 9 above)
Comment (mengjiao-liu) [reply cmd/kubeadm/app/kubeadm.go:27]: Yes, I need to look at this part of the code to see why it was added at the time.

**Comment by serathius** [comment]
Code freeze was lifted, @mengjiao-liu can you rebase this PR?

**Comment by mengjiao-liu** [comment]
All right. I'm going to rebase this PR. @serathius

File: staging/src/k8s.io/component-base/logs/logs.go [hunk 10]
```diff
@@ -31,9 +30,9 @@ const logFlushFreqFlagName = "log-flush-frequency"
 
 var logFlushFreq = pflag.Duration(logFlushFreqFlagName, 5*time.Second, "Maximum number of seconds between log flushes")
 
-func init() {
-	klog.InitFlags(flag.CommandLine)
-}
+//func init() {
```
Comment (serathius) [inline staging/src/k8s.io/component-base/logs/logs.go:33]: Let's not leave code commented out. Can you remove this?

**Comment by serathius** [comment]
/cc @mtaufen 
Can you take a look into component-base/config usage?

**Comment by mengjiao-liu** [comment]
Removing the log.init() method involves a lot of code and test code.
Therefore, for this PR, I have modified the way kubeadm uses flag to avoid `flag refined` error

**Comment by serathius** [comment]
/lgtm
PTAL @liggitt

File: cmd/kubeadm/app/cmd/cmd.go [hunk 11]
```diff
@@ -80,8 +83,14 @@ func NewKubeadmCommand(in io.Reader, out, err io.Writer) *cobra.Command {
 		},
 	}
 
-	cmds.ResetFlags()
+	namedFlagSets := &cliflag.NamedFlagSets{}
```
Comment (serathius) [inline cmd/kubeadm/app/cmd/cmd.go:86]: Sanity question, do we need to touch kubeadm at all. I don't see any place where they import `component-base/logs`.

I think we are going pretty deep into kubeadm, making harder to review. I would suggest thinking about if we can split this PR into smaller and ensure that the most important changes are prioritized. It might take multiple stages to get all changes in, but without this it will require root level approver which there are not many.

Comment (mengjiao-liu) [reply cmd/kubeadm/app/cmd/cmd.go:86]: > Sanity question, do we need to touch kubeadm at all. I don't see any place where they import component-base/logs.

In ` globalflag. AddGlobalFlags ` function  import `component-base/logs`
https://github.com/kubernetes/kubernetes/blob/2342ac3cfa46e4e69005fec6461012fdf9fc6dc3/staging/src/k8s.io/component-base/cli/globalflag/globalflags.go#L34

> I think we are going pretty deep into kubeadm, making harder to review. I would suggest thinking about if we can split this PR into smaller and ensure that the most important changes are prioritized. It might take multiple stages to get all changes in, but without this it will require root level approver which there are not many.

Yes, you are right, I will change it, try to change as little code as possible, make PR smaller.

**Comment by serathius** [comment]
/lgtm
PTAL @liggitt

**Comment by liggitt** [comment]
will defer the ack on the options/flags/config split to @deads2k or @sttts

File: cmd/kubeadm/app/cmd/options/generic.go [hunk 12]
```diff
@@ -100,3 +100,19 @@ func AddPatchesFlag(fs *pflag.FlagSet, patchesDir *string) {
 		`which patches are applied first alpha-numerically.`,
 	)
 }
+
+// SetGlobalFlags set global flags
+func SetGlobalFlags(fs *pflag.FlagSet) {
```
Comment (neolit123) [inline cmd/kubeadm/app/cmd/options/generic.go:105]: i don't see this function being called.
it duplicates cmd/kubeadm/app/kubeadm.go#Run()

Comment (mengjiao-liu) [reply cmd/kubeadm/app/cmd/options/generic.go]: Oh... This method is supposed to be added by using setGlobalFlags to set the klog parameter. Later, I did not want to change so much code, so I restored it again. However, I forgot to delete this part of code.

It's been delete.

File: cmd/kubeadm/app/kubeadm.go (hunk 9 above)
Comment (mengjiao-liu) [reply cmd/kubeadm/app/kubeadm.go]: > i think it's worth investigating when it was added (probably near the klog v2 migration?)
https://github.com/kubernetes/kubernetes/pull/70889/

This PR #70889 adds staging/src/k8s.io/component-base/logs/logs.go package klog.initFlags (Flag.mandline) code.
```
klog.initFlags (Flag.mandline)
```

And that's when cmd/kubeadm/kubeadm.go adds klog.InitFlag (nil) https://github.com/dims/kubernetes/blob/e558e291d1a41728da23f517e51b71038e3ba93e/cmd/kubeadm/kubeadm.go#L28

> see what can break if it's removed.

I tested what would happen if I removed the staging/src/k8s.io/component-base/logs/logs.go package init function, and there were a lot of errors. You can see that there are a lot of places that depend on the log package's klog.InitFlags(flag.CommandLine).
There are the tests logs:
https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/102326/pull-kubernetes-unit/1397473547823091712
https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/102326/pull-kubernetes-integration/1397482863271612416
https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/102326/pull-kubernetes-node-e2e-containerd/1397483544204283904/
https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/102326/pull-kubernetes-e2e-kind/1397483544095232000

Removing this method seems like a big job.

**Comment by serathius** [comment]
Is there any chance to get review on this? @sttts @dead2k
This is blocking a cleanup of logging flags.

File: pkg/kubelet/apis/config/validation/validation.go [hunk 13]
```diff
@@ -200,9 +200,9 @@ func ValidateKubeletConfiguration(kc *kubeletconfig.KubeletConfiguration) error
 	}
 	allErrors = append(allErrors, metrics.ValidateShowHiddenMetricsVersion(kc.ShowHiddenMetricsForVersion)...)
 
-	logOption := logs.NewOptions()
+	logOption := logsoptions.NewOptions()
 	if kc.Logging.Format != "" {
-		logOption.LogFormat = kc.Logging.Format
+		logOption.Config.Format = kc.Logging.Format
```
Comment (serathius) [inline pkg/kubelet/apis/config/validation/validation.go:205]: This was a workaround for defaulting, we can skip using NewOptions and instead call `ValidateLoggingConfiguration` directly on `kc.Logging`

**Comment by mengjiao-liu** [comment]
This PR is divided into smaller PRs #102462 #102431 for code review purposes.
/close
